### PR TITLE
REST API: corrections in swagger.json file generation

### LINKED ISSUE #2699: REST API: update swagger.json file
1. Now Swagger file which describes working on REST API contains a lot of redundant parameters.
**Clean up and leave only needed parameters with type "in: body".**
joxi_screenshot_1498730292474 1

Reference: #2675 
2. Parameters "X-User-Id" and "X-Auth-Token" is needless. The headers are used for Authorization and for Authorization Swagger has a special form. It means that a user should input values of User ID and Auth token into Authorization form not in headers parameter.
joxi_screenshot_1498732620127

**Remove parameters x_user_id and x_auth_token**

3. **Clean up code for Users endpoints implementation.**
Now it's desctibed as:
```js
ApiV1.swagger.meta.paths['/users'].get = { ... };
ApiV1.swagger.meta.paths['/users'].post = { .. };
ApiV1.swagger.meta.paths['/users/{id}'].get = { ... };
```
Should be:
```js
ApiV1.swagger.meta.paths = {
  '/users': {
    get: { ... },    
    post: { ... },
    ... 
  },
  '/users/{id}': {
    get: { ... },
  },
...
```

4. **Check that all endpoints with required authorization must have a lock sign and endpoints without authorization must not have a lock sign.**

For instance, POST method doesn't have a required authorization
joxi_screenshot_1498732972964

### LINKED ISSUE #2720: REST API: Swagger Response definitions to API and Organization methods
Definitions for responses to methods related to Api and Organization handling are missing from generated swagger document. 
Generating of them needs to added.

### PR DESCRIPTION
**Pull request by matleppa**
Closes #2699 
Closes #2720

### CODE REVIEW

**Comment by marla-singer** [comment]
@matleppa Well, I approve the first three commits, the code looks good. But the last one is really bad, looks like junk :confused: 

I will open the following branch where will try to upgrade solution for "$ref" parameters. Now PR is not yet ready for mergin

**Comment by matleppa** [comment]
@marla-singer OK. In case there were only one definition to be used in Users route, it would be more simple. However, when the functionality via API needs to follow the functionality via UI, it get messy, because in different situations the different sets of data is used in requests and responses. 
Also, when trying to use references instead of writing the definitions for single fields again and again, makes it more complicated, too.
I'm not happy with the outcome, but that was the only way to I could make it  work.
If you can find a clean and simple way to make it work, I'm more than happy to follow your lead.   ;-)

**Comment by marla-singer** [comment]
> in different situations the different sets of data is used in requests and responses.

Yes, I undesrtand that and I will keep in mind this case.

> Also, when trying to use references instead of writing the definitions for single fields again and again, makes it more complicated, too.

I'm not agree with you in this point but maybe I don't see enough but okay.

The first that confuses me is a lot of parameters which you described but never used or you described but I don't understand why?

```js
company: {
      name: 'company',
      in: 'body',
      description: 'Company name of user',
      required: true,
      type: 'string',
    },
    createdAt: {
      name: 'createdAt',
      in: 'body',
      description: 'Date and time in ISODate format, e.g. "2012-07-14T01:00:00+01:00" ',
      type: 'string',
    },
    email: {
      name: 'email',
      in: 'body',
      description: 'Email address for user',
      required: true,
      type: 'string',
    },
emailResponse: {
      properties: {
        name: 'address',
        in: 'body',
        description: 'Email address for user',
        type: 'string',
      },
    },
password: {
      name: 'password',
      in: 'body',
      description: 'Password for user',
      required: true,
      type: 'string',
    },
```

The property `in: body` means this parameter is used in Body request with POST method and there needs a schema of description.

For example, 
```js
userAddition: {
      name: 'user',
      in: 'body',
      description: 'Data for adding a new User',
      schema: {
        required: ['username', 'email', 'password'],
        properties: {
          username: {
            type: 'string',
            description: 'Username',
            example: 'johndoe',
          },
          email: {
            type: 'string',
            format: 'email',
            description: 'E-mail address of user',
            example: 'john.doe@ispname.com',
          },
          password: {
            type: 'string',
            description: 'Password for user',
            example: 'mypassword',
          },
        },
      },
    },
```

The second one, The description of a scheme should be in special part `"definition"`. I think this part was made special for that and if it will be unuseful and too complete to using, it wouldn't be made

**Comment by marla-singer** [comment]
Hm, I'm learning deeper your code and I start to understand the meaning of a lot of parameters. But My feeling is that it is not correctly using.

**Comment by matleppa** [comment]
Did you have a look at the swagger document, which is generated by your branch?
I think that the the references  i.e. "$ref: '#/definitions/loginResponse'" are not working, when used outside the original struct.

**Comment by marla-singer** [comment]
What's wrong?
joxi_screenshot_1499168646715

I changed the examples of authToken and userId properties for security. I'm not sure that the previous examples were good

**Comment by matleppa** [comment]
OK, now (after several reloads and meteor restarts) I get it working so that the references also work.

**Comment by matleppa** [comment]
@marla-singer I merged your branch here and now the params and definitions are in better shape. However, I still am not completely comfortable with the result. because in REST API functionality the lists of  input parameters are different in POST and PUT methods. (POST: username, email, password. PUT: username, password, company). 
I'm beginning to wonder, whether it would be easier to change the actual functionality such a way, that in both cases the all four parameters (username, email, password, company) could be given, although it is not possible via UI.

File: users/collection/server/api.js [hunk 1]
```diff
@@ -146,7 +146,7 @@ ApiV1.swagger.meta.paths = {
                 example: 'success',
               },
               data: {
-                $ref: '#/definitions/userPostResponse',
+                $ref: '#/definitions/userItem',
```
Comment (marla-singer) [inline users/collection/server/api.js]: @matleppa Good catch

**Comment by matleppa** [comment]
@marla-singer Sorry for increased content of this PR.

File: organizations/server/api.js [hunk 2]
```diff
@@ -28,7 +28,23 @@ ApiV1.addRoute('organizations', {
       ],
       responses: {
         200: {
+          name: 'organization',
+          in: 'body',
           description: 'Returns list of organizations',
```
Comment (marla-singer) [inline organizations/server/api.js]: Re-make this one

File: apis/server/api.js [hunk 3]
```diff
@@ -117,7 +133,20 @@ ApiV1.addCollection(Apis, {
         ],
         responses: {
           200: {
+            name: 'api',
+            in: 'body',
             description: 'Returns API',
```
Comment (marla-singer) [inline apis/server/api.js]: Re-make this one

File: apis/server/api.js [hunk 4]
```diff
@@ -138,7 +167,20 @@ ApiV1.addCollection(Apis, {
         ],
         responses: {
           200: {
+            name: 'api',
+            in: 'body',
             description: 'API successfully added',
```
Comment (marla-singer) [inline apis/server/api.js]: Re-make this one

File: apis/server/api.js [hunk 5]
```diff
@@ -204,7 +246,20 @@ ApiV1.addCollection(Apis, {
         ],
         responses: {
           200: {
+            name: 'api',
+            in: 'body',
             description: 'API successfully edited.',
```
Comment (marla-singer) [inline apis/server/api.js]: Re-make this one

File: core/server/api.js [hunk 6]
```diff
@@ -423,46 +473,49 @@ ApiV1.swagger = {
...
-            type: 'string',
-            example: 'sam.won@apinf.io',
-          },
-          verified: {
-            type: 'string',
-            example: 'false',
-          },
+    // Scheme for response parameters
+    // TODO: Response for API endpoints
+    // TODO: Response for Organization endpoints
```
Comment (marla-singer) [inline core/server/api.js]: Remove TODO comments

File: organizations/server/api.js [hunk 7]
```diff
@@ -99,7 +115,20 @@ ApiV1.addRoute('organizations', {
       ],
       responses: {
         200: {
+          name: 'organization',
+          in: 'body',
           description: 'Organization successfully added',
```
Comment (marla-singer) [inline organizations/server/api.js]: Re-make this one

File: organizations/server/api.js [hunk 8]
```diff
@@ -167,7 +196,16 @@ ApiV1.addRoute('organizations/:id', {
       ],
       responses: {
         200: {
-          description: 'Returns Organization',
+          name: 'organization',
+          in: 'body',
+          description: 'Returns organization',
```
Comment (marla-singer) [inline organizations/server/api.js]: Re-make this one

File: organizations/server/api.js [hunk 9]
```diff
@@ -212,7 +250,16 @@ ApiV1.addRoute('organizations/:id', {
       ],
       responses: {
         200: {
+          name: 'organization',
+          in: 'body',
           description: 'Organization successfully edited.',
```
Comment (marla-singer) [inline organizations/server/api.js]: Re-make this one

File: apis/server/api.js [hunk 10]
```diff
@@ -37,7 +37,23 @@ ApiV1.addCollection(Apis, {
         ],
         responses: {
           200: {
+            name: 'api',
+            in: 'body',
             description: 'Returns list of public APIs',
```
Comment (marla-singer) [inline apis/server/api.js]: @matleppa Incorrect structure for response. Must be:
```js
200: {
  description: "Description text",
  schema: {
    ... 
    data: {
      type: 'array',
      items: {
        $ref: "#/definitions/parameterName"
         }
      }
    }
 }
```

Respons structure doesn't need "name" and "in" property.
If `data` object has type as Array, items property should has "$ref" property.

Please re-make all incorrect responses

File: apis/server/api.js [hunk 11]
```diff
@@ -118,6 +133,17 @@ ApiV1.addCollection(Apis, {
...
             description: 'Returns API',
+            schema: {
+              type: 'object',
+              properties: {
+                status: {
+                  type: 'string',
+                  example: 'Success',
+                },
+                data:
+                  ApiV1.swagger.definitions.apiResponse,
```
Comment (marla-singer) [inline apis/server/api.js]: Re-write this using "$ref" parameter
```js
data: {
  $ref: "#definitions.apiResponse"
}
```

File: apis/server/api.js [hunk 12]
```diff
@@ -139,6 +165,17 @@ ApiV1.addCollection(Apis, {
...
             description: 'API successfully added',
+            schema: {
+              type: 'object',
+              properties: {
+                status: {
+                  type: 'string',
+                  example: 'Success',
+                },
+                data:
+                  ApiV1.swagger.definitions.apiResponse,
```
Comment (marla-singer) [inline apis/server/api.js]: Re-write this using "$ref" parameter
```js
data: {
  $ref: "#definitions.apiResponse"
}
```

File: apis/server/api.js [hunk 13]
```diff
@@ -205,6 +242,17 @@ ApiV1.addCollection(Apis, {
...
             description: 'API successfully edited.',
+            schema: {
+              type: 'object',
+              properties: {
+                status: {
+                  type: 'string',
+                  example: 'Success',
+                },
+                data:
+                  ApiV1.swagger.definitions.apiResponse,
```
Comment (marla-singer) [inline apis/server/api.js]: Re-write this using "$ref" parameter
```js
data: {
  $ref: "#definitions.apiResponse"
}
```

File: organizations/server/api.js [hunk 14]
```diff
@@ -100,6 +115,17 @@ ApiV1.addRoute('organizations', {
...
           description: 'Organization successfully added',
+          schema: {
+            type: 'object',
+            properties: {
+              status: {
+                type: 'string',
+                example: 'Success',
+              },
+              data:
+                ApiV1.swagger.definitions.organizationResponse,
```
Comment (marla-singer) [inline organizations/server/api.js]: Re-write this using "$ref" parameter
```js
data: {
  $ref: "#definitions.apiResponse"
}
```

File: organizations/server/api.js [hunk 15]
```diff
@@ -167,7 +193,14 @@ ApiV1.addRoute('organizations/:id', {
       ],
       responses: {
         200: {
-          description: 'Returns Organization',
+          description: 'Returns organization',
+          schema: {
+            type: 'object',
+            properties: {
+              data:
+                ApiV1.swagger.definitions.organizationResponse,
```
Comment (marla-singer) [inline organizations/server/api.js]: Re-write this using "$ref" parameter
```js
data: {
  $ref: "#definitions.apiResponse"
}
```

File: organizations/server/api.js [hunk 16]
```diff
@@ -213,6 +246,13 @@ ApiV1.addRoute('organizations/:id', {
       responses: {
         200: {
           description: 'Organization successfully edited.',
+          schema: {
+            type: 'object',
+            properties: {
+              data:
+                ApiV1.swagger.definitions.organizationResponse,
```
Comment (marla-singer) [inline organizations/server/api.js]: Re-write this using "$ref" parameter
```js
data: {
  $ref: "#definitions.apiResponse"
}
```

**Pull request merged**